### PR TITLE
feat: add relationship pagination

### PR DESCRIPTION
**Pull request by rbino**
# Contributor checklist

- [ ] Bug fixes include regression tests
- [ ] Chores
- [ ] Documentation changes
- [x] Features include unit/acceptance tests
- [ ] Refactoring
- [ ] Update dependencies

### CODE REVIEW

File: test/actions/load_test.exs [hunk 1]
```diff
@@ -1012,4 +1027,262 @@ defmodule Ash.Test.Actions.LoadTest do
...
+
+      assert %Ash.Page.Keyset{
+               results: [%{name: "category0"}, %{name: "category1"}]
+             } = post1.categories
+
+      assert %Ash.Page.Keyset{
+               results: [%{name: "category5"}, %{name: "category6"}]
+             } = post2.categories
+
+      # This doesn't work here but it works in ash_postgres
```
Comment (rbino) [inline test/actions/load_test.exs:1217]: cc @zachdaniel, probably a problem in the "lateral join emulation"

File: test/actions/load_test.exs [hunk 2]
```diff
@@ -1012,4 +1027,262 @@ defmodule Ash.Test.Actions.LoadTest do
...
+      #
+      # assert %Ash.Page.Keyset{
+      #          results: [%{name: "category2"}, %{name: "category3"}]
+      #        } = Ash.page!(post1.categories, :next)
+    end
+
+    test "works when nested" do
```
Comment (rbino) [inline test/actions/load_test.exs:1224]: This is currently not passing, it's my next step before undrafting this

**Comment by rbino** [comment]
This works correctly except for nesting (which I'm going to tackle next), but it doesn't work with count.

If I try to pass `count: true`, I get this stacktrace with `ash_postgres`:
```elixir
** (UndefinedFunctionError) function Tabemono.MenuManagement.DailyMenu.parameterized_type/2 is undefined or private
    (tabemono 0.1.0) Tabemono.MenuManagement.DailyMenu.parameterized_type(Ash.Type.UUID, [])
    (ash_sql 0.1.1-rc.11) lib/expr.ex:1738: AshSql.Expr.do_dynamic_expr/6
    (ash_sql 0.1.1-rc.11) lib/aggregate.ex:1504: AshSql.Aggregate.add_subquery_aggregate_select/6
    (ash_sql 0.1.1-rc.11) lib/aggregate_query.ex:179: anonymous fn/5 in AshSql.AggregateQuery.add_single_aggs/5
    (elixir 1.16.2) lib/enum.ex:2528: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ash_postgres 2.0.0-rc.9) lib/data_layer.ex:874: AshPostgres.DataLayer.run_aggregate_query_with_lateral_join/5
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:1748: Ash.Actions.Read.do_fetch_count/2
    (ash 3.0.0-rc.27) lib/ash/process_helpers.ex:38: anonymous fn/4 in Ash.ProcessHelpers.async/2
    (elixir 1.16.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.16.2) lib/task/supervised.ex:36: Task.Supervised.reply/4
    (ash 3.0.0-rc.27) lib/ash/process_helpers.ex:32: Ash.ProcessHelpers.async/2
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:1632: Ash.Actions.Read.fetch_count/4
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:419: anonymous fn/5 in Ash.Actions.Read.do_read/4
    (ash 3.0.0-rc.27) lib/ash/process_helpers.ex:38: anonymous fn/4 in Ash.ProcessHelpers.async/2
    (elixir 1.16.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.16.2) lib/task/supervised.ex:36: Task.Supervised.reply/4
    (ash 3.0.0-rc.27) lib/ash/process_helpers.ex:32: Ash.ProcessHelpers.async/2
    (ash 3.0.0-rc.27) lib/ash/process_helpers.ex:66: Ash.ProcessHelpers.task_with_timeout/5
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:559: Ash.Actions.Read.maybe_in_transaction/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:242: Ash.Actions.Read.do_run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:66: anonymous fn/3 in Ash.Actions.Read.run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:65: Ash.Actions.Read.run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/relationships.ex:365: anonymous fn/2 in Ash.Actions.Read.Relationships.do_fetch_related_records/3
    (elixir 1.16.2) lib/enum.ex:1708: anonymous fn/3 in Enum.map/2
    (elixir 1.16.2) lib/enum.ex:4396: anonymous fn/3 in Enum.map/2
    (elixir 1.16.2) lib/stream.ex:1816: anonymous fn/3 in Enumerable.Stream.reduce/3
    (elixir 1.16.2) lib/enum.ex:4839: Enumerable.List.reduce/3
    (elixir 1.16.2) lib/stream.ex:1828: Enumerable.Stream.do_each/4
    (elixir 1.16.2) lib/enum.ex:4396: Enum.map/2
    (ash 3.0.0-rc.27) lib/ash/actions/read/relationships.ex:40: Ash.Actions.Read.Relationships.fetch_related_records/2
    (ash 3.0.0-rc.27) lib/ash/actions/read/relationships.ex:24: Ash.Actions.Read.Relationships.load/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:263: Ash.Actions.Read.do_run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:66: anonymous fn/3 in Ash.Actions.Read.run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:65: Ash.Actions.Read.run/3
    (ash 3.0.0-rc.27) lib/ash.ex:1777: Ash.read/2
    (ash 3.0.0-rc.27) lib/ash.ex:1736: Ash.read!/2
    iex:30: (file)
```

In tests, it crashes since there's no `run_aggregate_query_with_lateral_join`:

```elixir
     ** (UndefinedFunctionError) function Ash.DataLayer.Ets.run_aggregate_query_with_lateral_join/5 is undefined or private. Did you mean:

           * run_aggregate_query/3
           * run_query_with_lateral_join/4
```

**Comment by rbino** [comment]
It seems that 9be1d2531459cb actually regresses a couple of tests, but if I revert it that there's another crash:

```elixir
** (Ecto.QueryError) Interpolated map values in :select can only be maps, structs, dynamics, subqueries and literals. Got [:daily_menu_dishes] in query:

from d0 in Tabemono.MenuManagement.Dish,
  as: 0,
  select: struct(d0, [:id, :name, :permanent, :inserted_at])

    (ecto 3.11.2) lib/ecto/query/builder/select.ex:278: Ecto.Query.Builder.Select.expand_nested/3
    (ecto 3.11.2) lib/ecto/query/builder/select.ex:290: Ecto.Query.Builder.Select.expand_nested_pair/3
    (elixir 1.16.2) lib/enum.ex:1832: anonymous fn/3 in Enum.map_reduce/3
    (stdlib 5.2.1) maps.erl:416: :maps.fold_1/4
    (elixir 1.16.2) lib/enum.ex:2540: Enum.map_reduce/3
    (ecto 3.11.2) lib/ecto/query/builder/select.ex:273: Ecto.Query.Builder.Select.expand_nested/3
    (ecto 3.11.2) lib/ecto/query/builder/select.ex:268: Ecto.Query.Builder.Select.expand_nested/3
    (ecto 3.11.2) lib/ecto/query/builder/select.ex:290: Ecto.Query.Builder.Select.expand_nested_pair/3
    (elixir 1.16.2) lib/enum.ex:1832: anonymous fn/3 in Enum.map_reduce/3
    (stdlib 5.2.1) maps.erl:416: :maps.fold_1/4
    (elixir 1.16.2) lib/enum.ex:2540: Enum.map_reduce/3
    (ecto 3.11.2) lib/ecto/query/builder/select.ex:273: Ecto.Query.Builder.Select.expand_nested/3
    (ecto 3.11.2) lib/ecto/query/builder/select.ex:290: Ecto.Query.Builder.Select.expand_nested_pair/3
    (elixir 1.16.2) lib/enum.ex:1832: anonymous fn/3 in Enum.map_reduce/3
    (stdlib 5.2.1) maps.erl:416: :maps.fold_1/4
    (elixir 1.16.2) lib/enum.ex:2540: Enum.map_reduce/3
    (ecto 3.11.2) lib/ecto/query/builder/select.ex:273: Ecto.Query.Builder.Select.expand_nested/3
    (ecto 3.11.2) lib/ecto/query/builder/select.ex:221: Ecto.Query.Builder.Select.select!/5
    (ash_sql 0.1.1-rc.11) lib/calculation.ex:102: AshSql.Calculation.add_calculations/5
    (ash 3.0.0-rc.27) lib/ash/query/query.ex:2865: Ash.Query.data_layer_query/2
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:436: anonymous fn/5 in Ash.Actions.Read.do_read/4
    (ash 3.0.0-rc.27) lib/ash/process_helpers.ex:38: anonymous fn/4 in Ash.ProcessHelpers.async/2
    (elixir 1.16.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.16.2) lib/task/supervised.ex:36: Task.Supervised.reply/4
    (ash 3.0.0-rc.27) lib/ash/process_helpers.ex:32: Ash.ProcessHelpers.async/2
    (ash 3.0.0-rc.27) lib/ash/process_helpers.ex:66: Ash.ProcessHelpers.task_with_timeout/5
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:568: Ash.Actions.Read.maybe_in_transaction/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:251: Ash.Actions.Read.do_run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:66: anonymous fn/3 in Ash.Actions.Read.run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:65: Ash.Actions.Read.run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/relationships.ex:365: anonymous fn/2 in Ash.Actions.Read.Relationships.do_fetch_related_records/3
    (elixir 1.16.2) lib/enum.ex:1708: anonymous fn/3 in Enum.map/2
    (elixir 1.16.2) lib/enum.ex:4396: anonymous fn/3 in Enum.map/2
    (elixir 1.16.2) lib/stream.ex:1816: anonymous fn/3 in Enumerable.Stream.reduce/3
    (elixir 1.16.2) lib/enum.ex:4839: Enumerable.List.reduce/3
    (elixir 1.16.2) lib/stream.ex:1828: Enumerable.Stream.do_each/4
    (elixir 1.16.2) lib/enum.ex:4396: Enum.map/2
    (ash 3.0.0-rc.27) lib/ash/actions/read/relationships.ex:40: Ash.Actions.Read.Relationships.fetch_related_records/2
    (ash 3.0.0-rc.27) lib/ash/actions/read/relationships.ex:24: Ash.Actions.Read.Relationships.load/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:272: Ash.Actions.Read.do_run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:66: anonymous fn/3 in Ash.Actions.Read.run/3
    (ash 3.0.0-rc.27) lib/ash/actions/read/read.ex:65: Ash.Actions.Read.run/3
    (ash 3.0.0-rc.27) lib/ash.ex:1777: Ash.read/2
    (ash 3.0.0-rc.27) lib/ash.ex:1736: Ash.read!/2
    iex:4: (file)
```

Which only happens with `ash_postgres` so it's not easily reproducible here. Strangely enough, this crash happens _only_ if I don't pass an explicit `offset` to pagination.

So

```elixir
dishes_load = Tabemono.MenuManagement.Dish |> Ash.Query.page(limit: 2, offset: 0)
[menu1, menu2] = DailyMenu |> Ash.read!(load: [dishes: dishes_load])
```

does not crash, while

```elixir
dishes_load = Tabemono.MenuManagement.Dish |> Ash.Query.page(limit: 2)
[menu1, menu2] = DailyMenu |> Ash.read!(load: [dishes: dishes_load])
```

does crash (with the stacktrace above)

File: lib/ash/actions/read/relationships.ex [hunk 3]
```diff
@@ -538,70 +538,51 @@ defmodule Ash.Actions.Read.Relationships do
...
+    } = unpaged
+
+    to_page_fun =
+      if relationship.cardinality == :many do
+        fn value, record ->
+          # We scope the lateral join to the specific record, so that next runs of rerun
+          # just fetch the entries related to this record
+          related_query =
+            Ash.Query.set_context(related_query, %{
+              data_layer: %{lateral_join_source: {[record], lateral_join_source_path}}
```
Comment (zachdaniel) [inline lib/ash/actions/read/relationships.ex:560]: clever.

**Comment by rbino** [comment]
I've ported the tests to `ash_postgres` using my local Ash branch and it seems that nested relationship pagination JustWorks :tm:  when tested there, so the failing test probably comes from the ETS Data Layers implementation of lateral joins, I'll check it

**Comment by rbino** [comment]
The first roadblock on pagination count was related to https://github.com/ash-project/ash_postgres/pull/251

Once that is removed, I now get an error which is similar to the one [here](https://github.com/ash-project/ash/pull/1050#issuecomment-2072954720):

```elixir
    ** (Ecto.SubQueryError) the following exception happened when compiling a subquery.

         ** (Ecto.QueryError) atoms, structs, maps, lists, tuples and sources are not allowed as map values in subquery, got: `%{__lateral_join_source__: &0}` in query:

[...]
```

**Comment by zachdaniel** [comment]
Interesting....this might be a problem. Can you make a repo that uses ash from this branch and latest postgres and has a failing test? I can look into it in the next few days.

**Comment by rbino** [comment]
I actually had a (dirty) fix locally that made this work: basically the problematic  bits seem to be the `select_merge` on `__lateral_join_source__` like [this](https://github.com/ash-project/ash_postgres/blob/5cfba7c6fa6f29a5311d7bdc052f97a97ba791c1/lib/data_layer.ex#L1021), if I temporarily remove them the count is returned. My idea was adding them conditionally based on an `aggregate?`flag passed to `lateral_join_query` but I was on my way out so I'll be able to look at this again on Monday.

_But_, and that would be the next problem I had on my list, even if after that hack the count returned correctly, the count logic needs to be restructured. Previously we always dealt with one paginated list at a time, now we have to return multiple counts, because if we are loading a paged relationship for N different root resources we need N different counts (and possibly we want to do it without doing N additional queries, so something like [this](https://stackoverflow.com/questions/61459505/get-count-of-related-rows-in-postgres-with-existing-lateral-join)). Right now the returned count is a single value and is the sum of all results.

File: test/actions/load_test.exs (hunk 1 above)
Comment (rbino) [reply test/actions/load_test.exs:1217]: [That](https://github.com/ash-project/ash/pull/1111) was it

**Comment by zachdaniel** [comment]
🤔 yeah this is going to be difficult. We need to support counting w/ grouping and limiting at the data layer. That is going to require new data layer callbacks likely 🥶

**Comment by zachdaniel** [comment]
Actually, I guess we could fetch many counts, like `count(where id == source_id) as count1, ....`? Still might be pretty hacky though...

**Comment by rbino** [comment]
Not sure if it might make sense to split this in two phases, i.e. support pagination on relationship _without support to count_ (erroring at runtime if `count: true` gets passed to a relationship) and then add the count in phase 2

**Comment by zachdaniel** [comment]
I think so, yes. But we should lay out what we think it will look like for said count support, just to make sure we don't lock ourselves out of it.

**Comment by zachdaniel** [comment]
So, I have some suspicions that there will be subtly breaking places when we use this, like things that expect relationships to come back as a list, but since this is opt-in, I don't see any reason we can't merge it, and tackle those as they come. Ready to go?

**Review by zachdaniel — APPROVED** [review]

**Comment by rbino** [comment]
Yes, I didn't find any obvious place where loaded relationship were used but when we found those places it's basically a matter of unwrapping results from the page (or return an error, if the stuff relationships are used for doesn't make sense when applied to a single page)

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

**Pull request merged**